### PR TITLE
Implement table asyncreadrows

### LINKED ISSUE #1218: Implement noex::Table::AsyncReadRows
Implement an operation to asynchronously read rows.  The API should look like this:

```C++
  template <
      typename RowFunctor, typename DoneFunctor, ...>
  std::shared_ptr<AsyncOperation> ReadRows(RowSet row_set, std::int64_t rows_limit, Filter filter,
      CompletionQueue& cq, RowFunctor&& row_callback, DoneFunctor&& done_callback) {
  }
```

`row_callback` should be called each time a new row is ready, and should meet this requirement:

```C++
static_assert(
    google::cloud::internal::is_invocable<
              RowFunctor, CompletionQueue&, Row&, grpc::Status&>
```

`done_callback` should be called when all the rows are read (or the request fails), and should meet this requirement:

```C++
static_assert(
    google::cloud::internal::is_invocable<
              DoneFunctor, CompletionQueue&, grpc::Status&>
```

The returned `AsyncOperation` can be used to cancel the iteration.

### PR DESCRIPTION
**Pull request by manish-qlogic**
This is an attempt to fix #1218

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1587)

### CODE REVIEW

File: google/cloud/bigtable/internal/async_row_reader.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+ private:
+  template <typename Functor,
+            typename std::enable_if<
+                google::cloud::internal::is_invocable<Functor, CompletionQueue&,
+                                                      grpc::Status&>::value,
+                int>::type valid_callback_type = 0>
+  struct FinishedCallback {
+    FinishedCallback(AsyncRowReader& parent, Functor&& callback)
+        : parent_(parent), callback_(callback) {}
```
Comment (devjgm) [inline google/cloud/bigtable/internal/async_row_reader.h]: Mabye `std::move(callback)`?

File: google/cloud/bigtable/internal/table.h [hunk 2]
```diff
@@ -255,6 +256,53 @@ class Table {
...
+   *
+   * @throws std::runtime_error if rows_limit is < 0. rows_limit = 0(default)
+   * will return all rows
+   * @param cq the completion queue that will execute the asynchronous calls,
+   *     the application must ensure that one or more threads are blocked on
+   *     `cq.Run()`.
+   * @param callback a functor to be called when the row is ready for reading.
+   * @param callback a functor to be called when the operation completes. It
+   *     must satisfy (using C++17 types):
+   *     static_assert(std::is_invocable< Functor, CompletionQueue&,
```
Comment (devjgm) [inline google/cloud/bigtable/internal/table.h]: Is the space at `< Functor` intentional, or a typo?

File: google/cloud/bigtable/testing/inprocess_data_client.cc [hunk 3]
```diff
@@ -69,6 +69,14 @@ InProcessDataClient::ReadRows(grpc::ClientContext* context,
   return Stub()->ReadRows(context, request);
 }
 
+std::unique_ptr<::grpc::ClientAsyncReaderInterface<btproto::ReadRowsResponse>>
+InProcessDataClient::AsyncReadRows(
+    ::grpc::ClientContext* context,
+    const ::google::bigtable::v2::ReadRowsRequest& request,
```
Comment (devjgm) [inline google/cloud/bigtable/testing/inprocess_data_client.cc]: I see a mix of west-const and east-const. West-const here, and east const in the signature 7 lines above. I tend to prefer west-const in general for reasons ;-), but I mostly prefer consistency. May want to change this to east-const to match the surrounding code more, or change other cases to west-const. I'm not really placing a vote here, just an observation.

**Comment by manish-qlogic** [comment]
PTAL

**Comment by coryan** [comment]
You also need to rebase to pick the Bazel fixes.

**Comment by coryan** [comment]
@manish-qlogic Can you do another rebase?  We added a macOS build on Kokoro and you need to pick those changes too.

**Comment by manish-qlogic** [comment]
Rebased.

**Comment by coryan** [comment]
I just merged another change that conflicts with this. I am afraid another rebase is needed.

**Comment by manish-qlogic** [comment]
Rebased.

**Pull request merged**